### ISSUE TITLE
DQN的一些改进建议

### ISSUE THREAD

**Issue by Hyperion-shuo** [7eu7d7/genshin_auto_fish#69]
1. 我看您在环境中定义了done，但是在构建TD target 的时候没有使用，严谨的用法应该将done也存储在replay buffer中，
    并在构建TD target的时候加上。
   `q_target = b_r + self.gamma * （1 - d）* q_next.max(dim=1)[0].view(self.batch_size, 1)`

2. Double DQN相较于DQN的提升还是很可观的，实现也只需要额外加一两行代码。只需额外forward一次，使用eval_net得到argmax a'，用该a'去q_next中选择动作构建TD target，比如:
 `q_eval_next = self.eval_net(b_s_).detach() `
 `b_a_ = torch.argmax(q_eval_next, dim=1, keepdim=True)    # get eval_net's argmax_a'(Q(s', a')) `
 `q_target_next = self.target_net(b_s_).detach()   `
 `q_target = b_r + self.gamma * (1 - d) * q_target_next.gather(1, b_a_)`

3.  epsilon-greedy中的epsilon通常是以epsilon的概率去探索，您这边反过来了，以（1-epsilon）探索，可能会产生歧义。

4. 关于DQN冷启动的问题，一开始策略比较差且replay_buffer为空时，可以先用random policy跑几个episode把数据存到replay_buffer中，这样后面学的更快。

5. 不对的地方大佬轻喷，逃～

**Comment by 7eu7d7** [comment]
我在训练过程中done的这一个我应该也加进去了，我是先添加再最后才break的。

epsilon-greedy这个我用的是小于epsilon则使用eval_net的结果，也就是说0.1的概率随机选取，应该是没问题的，可能我符号写的不一样。

**Comment by 7eu7d7** [comment]
试了一下Double DQN，泛化能力提升确实挺明显，只在仿真环境训练的放原神里面都不错了

**Comment by Hyperion-shuo** [comment]
> 我在训练过程中done的这一个我应该也加进去了，我是先添加再最后才break的。
> 
> epsilon-greedy这个我用的是小于epsilon则使用eval_net的结果，也就是说0.1的概率随机选取，应该是没问题的，可能我符号写的不一样。

我的意思是您这边代码没问题，只是强化学习里epsilon惯例里是使用epsilon去探索

**Comment by Hyperion-shuo** [comment]
> 试了一下Double DQN，泛化能力提升确实挺明显，只在仿真环境训练的放原神里面都不错了

有效果就好，老哥动手能力是真强
